sql/handler.cc

```diff
@@ -2658,9 +2660,12 @@ static my_bool xarecover_handlerton(THD *unused, plugin_ref plugin,
         */
         if (info->mem_root)
         {
-          // remember "full" xid too when it's not in mysql format
+          // remember "full" xid too when it's not in mysql format.
+          // MDEV-27342 - Also record the server_id of the transaction. It will be used for
+          // populating the input XID to be searched in hash.
           if (xid_member_replace(info->commit_list, x, info->mem_root,
-                                 is_server_xid? NULL : &info->list[i]))
+                                 is_server_xid? NULL : &info->list[i],
+                                 is_server_xid? info->list[i].get_trx_server_id(): server_id))
```

Comment:
Right, we have to pass the xid's `server_id` down to  `xid_recovery_member` ctor, but I suggest to
pass zero (which would be the default of  `xid_recovery_member::server_id`) when `get_trx_server_id() == server_id`.
As the latter is the normal (It feels not so many users recover along your way :-))  case,  `xid_recovery_member::server_id == 0` would avoid `memcpy` in `xid_t::set`.

Reply:
Thank you! Good point! Updated in new commit.